Tests/Realm.Tests/Sync/SessionTests.cs

```diff
@@ -270,31 +237,388 @@ public void Session_ClientReset_DiscardLocal_ManualResetFallback_InitiateClientR
+        public void Session_ClientResetHandlers_OnBefore_And_OnAfter(IClientResetHandler handler)
         {
-            const string validValue = "this will sync";
-            const string invalidValue = "this will be deleted";
-
             SyncTestHelpers.RunBaasTestAsync(async () =>
             {
-                var tcs = new TaskCompletionSource<object>();
                 var onBeforeTriggered = false;
+                var onAfterTriggered = false;
+                var tcs = new TaskCompletionSource<object>();
                 var config = await GetIntegrationConfigAsync();
-                config.ClientResetHandler = new DiscardLocalResetHandler
+
+                var beforeCb = GetOnBeforeHandler(tcs, beforeFrozen =>
+                {
+                    Assert.That(onBeforeTriggered, Is.False);
+                    Assert.That(onAfterTriggered, Is.False);
+                    onBeforeTriggered = true;
+                });
+
+                var afterCb = GetOnAfterHandler(tcs, (beforeFrozen, after) =>
+                {
+                    Assert.That(onBeforeTriggered, Is.True);
+                    Assert.That(onAfterTriggered, Is.False);
+                    onAfterTriggered = true;
+                });
+
+                config.ClientResetHandler = handler.BuildHandler(beforeCb, afterCb);
+
+                using var realm = await GetRealmAsync(config);
+
+                GetSession(realm).SimulateClientReset("simulated client reset");
+
+                await tcs.Task;
+
+                Assert.That(onBeforeTriggered, Is.True);
+                Assert.That(onAfterTriggered, Is.True);
+            });
+        }
+
+        /* If any ArrayDelete or ArrayMove operates on indices not added by the recovery, such operation will be discarded.
+         * Hence, the last client to experience a reset will "win" the array state.
+         *
+         * 1. clientA adds objectA with array innerObj[0,1,2] and syncs it, then disconnects
+         * 2. clientB starts and syncs the same objectA, then disconnects
+         * 3. While offline, clientA deletes innerObj[2] while clientB swaps innerObj[0] with innerObj[2]
+         * 4. A client reset is triggered on the server
+         * 5. clientA goes online and uploads the changes
+         * 6. only now clientB goes online, downloads and merges the changes. Only innerObj[0,1] exist and the swap is discarded
+         */
+        [Test]
+        public void SessionIntegrationTest_ClientResetHandlers_OutOfBoundArrayMove_Discarded()
+        {
+            SyncTestHelpers.RunBaasTestAsync(async () =>
+            {
+                var user = await GetUserAsync();
+                var partition = Guid.NewGuid().ToString();
+
+                // ===== clientA =====
+                var tscAfterClientResetMergeA = new TaskCompletionSource<object>();
+                var configA = await GetIntegrationConfigAsync(partition, null, "realm123.realm", user);
```

Comment:
It's a bad practice to pass `null` as a positional argument as it's not clear what it means. Since that is an optional argument, it's a better idea to name the remaining arguments. Also, let's not specify the path of the realm explicitly.

```suggestion
                var configA = await GetIntegrationConfigAsync(partition, user: user);
```